File: scripts/auto_commit.py
#!/usr/bin/env python3
"""
Cognitive Ergonomics Auto-Commit Script

Implements Cognitive Ergonomics Framework principles for git workflows:
- Miller's Law (7±2 working memory limit)
- Hick-Hyman Law (minimize decision points)
- Gestalt Grouping (logical change clustering)
- Progressive Disclosure (information layering)
- Flow State Preservation (minimize context switching)

Automatically generates ergonomic commit messages, validates against cognitive
load constraints, and provides optimized push/PR workflows.
"""

import argparse
import json
import subprocess
import sys
from pathlib import Path
from typing import Dict, List, Optional, Tuple


# Commit message format from config/hooks.json
# Format: ^(feat|fix|docs|style|refactor|test|chore|quantum)\(\w+\): .{1,72}
# Hierarchical commit type grouping (Hick-Hyman Law optimization)
# Grouped into 3 primary + 5 secondary to minimize decision load
PRIMARY_COMMIT_TYPES = ["feat", "fix", "chore"]
SECONDARY_COMMIT_TYPES = ["docs", "style", "refactor", "test", "quantum"]
COMMIT_TYPES = PRIMARY_COMMIT_TYPES + SECONDARY_COMMIT_TYPES


def run_git(args: List[str], repo_path: str = ".") -> Tuple[bool, str]:
    """Run a git command and return success status and output."""
    try:
        result = subprocess.run(
            ["git"] + args,
            cwd=Path(repo_path).resolve(),
            capture_output=True,
            text=True,
            check=True
        )
        return True, result.stdout.strip()
    except subprocess.CalledProcessError as e:
        return False, e.stderr.strip()


def get_current_branch(repo_path: str = ".") -> str:
    """Get the current branch name."""
    success, output = run_git(["rev-parse", "--abbrev-ref", "HEAD"], repo_path)
    return output if success else "unknown"


def get_staged_files(repo_path: str = ".") -> List[str]:
    """Get list of staged files."""
    success, output = run_git(["diff", "--cached", "--name-only"], repo_path)
    if not success or not output:
        return []
    return [f.strip() for f in output.split('\n') if f.strip()]


def get_staged_changes(repo_path: str = ".") -> Dict[str, int]:
    """Get statistics about staged changes."""
    success, output = run_git(["diff", "--cached", "--stat"], repo_path)
    if not success or not output:
        return {"files": 0, "insertions": 0, "deletions": 0}
    
    # Parse the last line which contains the summary
    lines = output.strip().split('\n')
    if not lines:
        return {"files": 0, "insertions": 0, "deletions": 0}
    
    summary = lines[-1]
    stats = {"files": 0, "insertions": 0, "deletions": 0}
    
    # Parse summary like "3 files changed, 150 insertions(+), 10 deletions(-)"
    parts = summary.split(',')
    for part in parts:
        part = part.strip()
        if 'file' in part:
            try:
                stats["files"] = int(part.split()[0])
            except (ValueError, IndexError):
                pass
        elif 'insertion' in part:
            try:
                stats["insertions"] = int(part.split()[0])
            except (ValueError, IndexError):
                pass
        elif 'deletion' in part:
            try:
                stats["deletions"] = int(part.split()[0])
            except (ValueError, IndexError):
                pass
    
    return stats


def analyze_file_changes(staged_files: List[str]) -> Dict[str, List[str]]:
    """Categorize file changes by type."""
    categories = {
        "source": [],
        "test": [],
        "docs": [],
        "config": [],
        "build": [],
        "other": []
    }
    
    for file in staged_files:
        file_lower = file.lower()
        path = Path(file)
        
        # Categorize based on file path and extension
        if any(test_dir in file_lower for test_dir in ["test", "tests", "_test.", "spec"]):
            categories["test"].append(file)
        elif any(doc_dir in file_lower for doc_dir in ["doc", "docs", "readme", "changelog"]):
            categories["docs"].append(file)
        elif any(config_ext in path.suffix.lower() for config_ext in [".json", ".yaml", ".yml", ".toml", ".ini", ".cfg"]):
            categories["config"].append(file)
        elif any(build_file in file_lower for build_file in ["makefile", "cmake", "dockerfile", "docker-compose", ".github/workflows"]):
            categories["build"].append(file)
        elif any(source_ext in path.suffix.lower() for source_ext in [".py", ".cpp", ".c", ".h", ".hpp", ".go", ".rs", ".js", ".ts"]):
            categories["source"].append(file)
        else:
            categories["other"].append(file)
    
    return categories


def determine_commit_type(categories: Dict[str, List[str]], branch_name: str) -> str:
    """Determine the commit type based on file categories and branch name."""
    # Check branch name first
    if branch_name.startswith("feature/"):
        return "feat"
    elif branch_name.startswith("fix/"):
        return "fix"
    elif branch_name.startswith("docs/"):
        return "docs"
    elif branch_name.startswith("refactor/"):
        return "refactor"
    elif branch_name.startswith("test/"):
        return "test"
    
    # Check file categories
    if categories["test"] and not categories["source"]:
        return "test"
    elif categories["docs"] and not categories["source"]:
        return "docs"
    elif categories["config"] and not categories["source"]:
        return "chore"
    elif any("quantum" in f.lower() for f in categories["source"]):
        return "quantum"
    elif categories["source"]:
        # Check if it's a new feature or a fix
        # For now, default to feat for source changes
        return "feat"
    
    # Default
    return "chore"


def generate_scope(categories: Dict[str, List[str]]) -> str:
    """
    Generate commit scope following Miller's Law (max 2 scopes).
    
    Cognitive Principle: Human working memory can only hold 7±2 items.
    By limiting scope to maximum 2 identifiers, we minimize cognitive load
    and ensure commit messages are rapidly parsable.
    """
    scopes = []
    
    # Check for specific modules or components
    all_files = []
    for file_list in categories.values():
        all_files.extend(file_list)
    
    # Look for common patterns in file paths
    for file in all_files:
        path = Path(file)
        parts = path.parts
        
        # Check for module names
        for part in parts:
            if part in ["quantum_core", "dll_bridge", "go_runtime", "sycl_accelerator", 
                       "runtime_engine", "rag_service", "test_orchestrator", "wui_designer"]:
                if part not in scopes:
                    scopes.append(part)
        
        # Check for specific file types
        if "hooks" in file.lower() and "hooks" not in scopes:
            scopes.append("hooks")
        if "config" in file.lower() and "config" not in scopes:
            scopes.append("config")
    
    # STRICT: Limit to maximum 2 scopes (Miller's Law optimization)
    if len(scopes) > 2:
        scopes = scopes[:2]
    
    return ", ".join(scopes) if scopes else ""


def generate_commit_message(commit_type: str, scope: str, stats: Dict[str, int], 
                          categories: Dict[str, List[str]]) -> str:
    """Generate a commit message following the project's format."""
    # Generate a descriptive message
    if stats["files"] == 1:
        file_desc = f"update {Path(categories['source'][0] if categories['source'] else categories['other'][0]).name}"
    else:
        file_desc = f"{stats['files']} files"
    
    # Create message based on type and changes
    if commit_type == "feat":
        if stats["files"] == 1 and categories["source"]:
            message = f"add {Path(categories['source'][0]).stem} functionality"
        else:
            message = f"add {file_desc} with {stats['insertions']} insertions"
    elif commit_type == "fix":
        message = f"fix issues in {file_desc}"
    elif commit_type == "docs":
        message = f"update documentation in {file_desc}"
    elif commit_type == "refactor":
        message = f"refactor {file_desc}"
    elif commit_type == "test":
        message = f"add/update tests for {file_desc}"
    elif commit_type == "chore":
        message = f"update {file_desc}"
    elif commit_type == "quantum":
        message = f"update quantum operations in {file_desc}"
    else:
        message = f"update {file_desc}"
    
    # Format according to project convention: type(scope): message
    if scope:
        return f"{commit_type}({scope}): {message}"
    else:
        return f"{commit_type}: {message}"


def validate_cognitive_constraints(categories: Dict[str, List[str]], stats: Dict[str, int]) -> Tuple[bool, List[str]]:
    """
    Validate commit against Cognitive Ergonomics Framework constraints.
    
    Checks:
    - Change count does not exceed working memory limits
    - Changes are logically grouped (Gestalt principle)
    - Single logical purpose per commit
    """
    warnings = []
    valid = True
    
    # Miller's Law: >15 files exceeds cognitive parsing capacity
    if stats["files"] > 15:
        warnings.append(f"⚠️  COGNITIVE WARNING: {stats['files']} files changed. Consider splitting into smaller atomic commits.")
        valid = False
    
    # Gestalt Grouping: multiple unrelated change types indicate unchunked work
    change_types = sum(1 for files in categories.values() if len(files) > 0)
    if change_types > 3:
        warnings.append("⚠️  COGNITIVE WARNING: Multiple unrelated change types detected. This commit spans multiple domains.")
        valid = False
    
    # Flow State: very large changes break review continuity
    if stats["insertions"] + stats["deletions"] > 500:
        warnings.append(f"⚠️  COGNITIVE WARNING: {stats['insertions'] + stats['deletions']} total lines changed. Large diffs reduce review quality.")
    
    return valid, warnings


def validate_commit_message(message: str) -> Tuple[bool, str]:
    """Validate commit message format AND cognitive ergonomics constraints."""
    import re
    pattern = r"^(feat|fix|docs|style|refactor|test|chore|quantum)(\(\w+\))?: .{1,72}$"
    
    if re.match(pattern, message):
        # Additional ergonomic validation
        if len(message) > 60:
            return True, "Commit message valid (optimal length for rapid scanning)"
        return True, "Commit message format is valid"
    else:
        return False, f"Commit message does not match required format: {pattern}"


def auto_commit(repo_path: str = ".", auto_push: bool = False, 
               create_pr: bool = False, dry_run: bool = False) -> Dict:
    """
    Automatically commit staged changes with a generated commit message.
    
    Args:
        repo_path: Path to the git repository
        auto_push: Whether to push after committing
        create_pr: Whether to create a PR after pushing
        dry_run: If True, only generate the message without committing
    
    Returns:
        Dictionary with results
    """
    result = {
        "success": False,
        "commit_message": "",
        "staged_files": [],
        "stats": {},
        "error": None,
        "dry_run": dry_run,
        "pr_created": False,
        "pr_url": None
    }
    
    # Check if we're in a git repository
    success, _ = run_git(["status"], repo_path)
    if not success:
        result["error"] = "Not in a git repository"
        return result
    
    # Get current branch
    branch_name = get_current_branch(repo_path)
    if branch_name == "unknown":
        result["error"] = "Could not determine current branch"
        return result
    
    # Get staged files
    staged_files = get_staged_files(repo_path)
    if not staged_files:
        result["error"] = "No staged files to commit"
        return result
    
    result["staged_files"] = staged_files
    
    # Get change statistics
    stats = get_staged_changes(repo_path)
    result["stats"] = stats
    
    # Analyze file changes
    categories = analyze_file_changes(staged_files)
    
    # Determine commit type and scope
    commit_type = determine_commit_type(categories, branch_name)
    scope = generate_scope(categories)
    
    # Generate commit message
    commit_message = generate_commit_message(commit_type, scope, stats, categories)
    result["commit_message"] = commit_message
    
    # Validate commit message
    is_valid, validation_message = validate_commit_message(commit_message)
    if not is_valid:
        result["error"] = f"Generated commit message is invalid: {validation_message}"
        return result
    
    # Apply Cognitive Ergonomics validation
    cognitive_valid, cognitive_warnings = validate_cognitive_constraints(categories, stats)
    result["cognitive_warnings"] = cognitive_warnings
    
    # Progressive Disclosure: Only show detailed stats when requested or warnings exist
    if cognitive_warnings:
        print("\n" + "═"*60)
        print("  📊 COGNITIVE ERGONOMICS FEEDBACK")
        print("═"*60)
        for warning in cognitive_warnings:
            print(f"  {warning}")
        print("═"*60 + "\n")
    
    if dry_run:
        result["success"] = True
        return result
    
    # Create the commit
    success, output = run_git(["commit", "-m", commit_message], repo_path)
    if not success:
        result["error"] = f"Failed to create commit: {output}"
        return result
    
    # Push if requested
    if auto_push:
        success, output = run_git(["push", "-u", "origin", branch_name], repo_path)
        if not success:
            result["error"] = f"Failed to push: {output}"
            return result
    
    # Create PR if requested
    if create_pr and auto_push:
        try:
            # Try to use GitHub CLI to create PR
            import subprocess
            from datetime import datetime
            
            # Generate PR title from commit message
            pr_title = commit_message
            
            # Generate PR body
            pr_body = f"""## Auto-Generated PR

This PR was automatically created by the Kanban Auto-Commit workflow.

**Commit Message:** {commit_message}

### Changes
- Auto-committed staged changes
- Generated commit message following project conventions

---
*Generated on {datetime.now().isoformat()}*
"""
            
            # Try to create PR using GitHub CLI
            gh_result = subprocess.run(
                ["gh", "pr", "create",
                 "--title", pr_title,
                 "--body", pr_body,
                 "--base", "develop",
                 "--head", branch_name],
                capture_output=True,
                text=True
            )
            
            if gh_result.returncode == 0:
                # Extract PR URL from output
                pr_url = gh_result.stdout.strip()
                result["pr_created"] = True
                result["pr_url"] = pr_url
                print(f"PR created successfully: {pr_url}")
            else:
                # Fall back to manual PR creation message
                print(f"Could not create PR automatically: {gh_result.stderr}")
                print("Please create PR manually with the following details:")
                print(f"Title: {pr_title}")
                print(f"Base: develop")
                print(f"Head: {branch_name}")
                print(f"Body: {pr_body[:200]}...")
                
        except Exception as e:
            print(f"Error creating PR: {e}")
            print("Please create PR manually")
    
    result["success"] = True
    return result


def enable_unicode_support():
    """Enable Unicode output support for Windows terminals."""
    import sys
    if sys.platform == 'win32':
        # Set UTF-8 encoding for stdout/stderr on Windows
        sys.stdout.reconfigure(encoding='utf-8')
        sys.stderr.reconfigure(encoding='utf-8')


def main():
    enable_unicode_support()
    
    parser = argparse.ArgumentParser(
        description="Auto-commit staged changes with generated commit messages"
    )
    parser.add_argument('--repo-path', default='.', help='Path to git repository')
    parser.add_argument('--auto-push', action='store_true', help='Push after committing')
    parser.add_argument('--create-pr', action='store_true', help='Create PR after pushing')
    parser.add_argument('--dry-run', action='store_true', help='Only generate message, don\'t commit')
    parser.add_argument('--json', action='store_true', help='Output result as JSON')
    
    args = parser.parse_args()
    
    result = auto_commit(
        repo_path=args.repo_path,
        auto_push=args.auto_push,
        create_pr=args.create_pr,
        dry_run=args.dry_run
    )
    
    if args.json:
        print(json.dumps(result, indent=2))
    else:
        if result["success"]:
            if result["dry_run"]:
                # Fitts's Law optimized layout: critical information first, grouped vertically
                print("\n" + "═"*60)
                print("  ✅  COMMIT PREVIEW (DRY RUN)")
                print("═"*60)
                print(f"\n      📝  {result['commit_message']}")
                print(f"\n      📁  {len(result['staged_files'])} files changed")
                print(f"      📊  +{result['stats']['insertions']} insertions / -{result['stats']['deletions']} deletions")
                print("\n" + "═"*60 + "\n")
            else:
                print("\n" + "═"*60)
                print("  ✅  COMMIT SUCCESSFUL")
                print("═"*60)
                print(f"\n      📝  {result['commit_message']}")
                print("\n" + "═"*60)
                
                if args.auto_push:
                    print("\n      🚀  Pushing changes to remote...")
                    print("      ✓  Changes pushed successfully")
                    print()
        else:
            print("\n" + "═"*60)
            print("  ❌  COMMIT FAILED")
            print("═"*60)
            print(f"\n      {result['error']}")
            print("\n" + "═"*60 + "\n", file=sys.stderr)
            sys.exit(1)


if __name__ == '__main__':
    main()

File: go_cli/cmd/auto-commit/auto_commit.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var commitTypes = []string{"feat", "fix", "docs", "style", "refactor", "test", "chore", "quantum"}

// CommitResult contains result of auto commit operation
type CommitResult struct {
	Success       bool              `json:"success"`
	CommitMessage string            `json:"commit_message"`
	StagedFiles   []string          `json:"staged_files"`
	Stats         map[string]int    `json:"stats"`
	Error         string            `json:"error,omitempty"`
	DryRun        bool              `json:"dry_run"`
	PRCreated     bool              `json:"pr_created"`
	PRURL         string            `json:"pr_url,omitempty"`
}

// AutoCommit is the main auto commit handler
type AutoCommit struct {
	RepoPath string
}

// NewAutoCommit creates a new AutoCommit instance
func NewAutoCommit(repoPath string) *AutoCommit {
	if repoPath == "" {
		repoPath = "."
	}
	absPath, _ := filepath.Abs(repoPath)
	return &AutoCommit{
		RepoPath: absPath,
	}
}

// RunGit executes a git command
func (ac *AutoCommit) RunGit(args ...string) (bool, string) {
	cmd := exec.Command("git", args...)
	cmd.Dir = ac.RepoPath

	var out bytes.Buffer
	var stderr bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &stderr

	err := cmd.Run()
	if err != nil {
		return false, stderr.String()
	}

	return true, strings.TrimSpace(out.String())
}

// GetCurrentBranch returns current git branch name
func (ac *AutoCommit) GetCurrentBranch() string {
	success, output := ac.RunGit("rev-parse", "--abbrev-ref", "HEAD")
	if !success {
		return "unknown"
	}
	return output
}

// GetStagedFiles returns list of staged files
func (ac *AutoCommit) GetStagedFiles() []string {
	success, output := ac.RunGit("diff", "--cached", "--name-only")
	if !success || output == "" {
		return []string{}
	}

	files := strings.Split(output, "\n")
	var result []string
	for _, f := range files {
		f = strings.TrimSpace(f)
		if f != "" {
			result = append(result, f)
		}
	}
	return result
}

// GetStagedChanges returns statistics about staged changes
func (ac *AutoCommit) GetStagedChanges() map[string]int {
	stats := map[string]int{
		"files": 0, "insertions": 0, "deletions": 0,
	}

	success, output := ac.RunGit("diff", "--cached", "--stat")
	if !success || output == "" {
		return stats
	}

	lines := strings.Split(output, "\n")
	if len(lines) == 0 {
		return stats
	}

	summary := lines[len(lines)-1]
	parts := strings.Split(summary, ",")

	for _, part := range parts {
		part = strings.TrimSpace(part)
		fields := strings.Fields(part)
		if len(fields) < 2 {
			continue
		}

		val, err := strconv.Atoi(fields[0])
		if err != nil {
			continue
		}

		if strings.Contains(part, "file") {
			stats["files"] = val
		} else if strings.Contains(part, "insertion") {
			stats["insertions"] = val
		} else if strings.Contains(part, "deletion") {
			stats["deletions"] = val
		}
	}

	return stats
}

// AnalyzeFileChanges categorizes file changes by type
func (ac *AutoCommit) AnalyzeFileChanges(stagedFiles []string) map[string][]string {
	categories := map[string][]string{
		"source":  {}, "test": {}, "docs": {}, "config": {}, "build": {}, "other": {},
	}

	testPatterns := []string{"test", "tests", "_test.", "spec"}
	docPatterns := []string{"doc", "docs", "readme", "changelog"}
	configExt := []string{".json", ".yaml", ".yml", ".toml", ".ini", ".cfg"}
	buildPatterns := []string{"makefile", "cmake", "dockerfile", "docker-compose", ".github/workflows"}
	sourceExt := []string{".py", ".cpp", ".c", ".h", ".hpp", ".go", ".rs", ".js", ".ts"}

	for _, file := range stagedFiles {
		fileLower := strings.ToLower(file)
		ext := filepath.Ext(file)

		classified := false

		// Test files
		for _, p := range testPatterns {
			if strings.Contains(fileLower, p) {
				categories["test"] = append(categories["test"], file)
				classified = true
				break
			}
		}
		if classified {
			continue
		}

		// Docs
		for _, p := range docPatterns {
			if strings.Contains(fileLower, p) {
				categories["docs"] = append(categories["docs"], file)
				classified = true
				break
			}
		}
		if classified {
			continue
		}

		// Config
		for _, e := range configExt {
			if ext == e {
				categories["config"] = append(categories["config"], file)
				classified = true
				break
			}
		}
		if classified {
			continue
		}

		// Build
		for _, p := range buildPatterns {
			if strings.Contains(fileLower, p) {
				categories["build"] = append(categories["build"], file)
				classified = true
				break
			}
		}
		if classified {
			continue
		}

		// Source
		for _, e := range sourceExt {
			if ext == e {
				categories["source"] = append(categories["source"], file)
				classified = true
				break
			}
		}
		if classified {
			continue
		}

		// Other
		categories["other"] = append(categories["other"], file)
	}

	return categories
}

// DetermineCommitType identifies commit type from changes
func (ac *AutoCommit) DetermineCommitType(categories map[string][]string, branchName string) string {
	// Check branch name first
	switch {
	case strings.HasPrefix(branchName, "feature/"):
		return "feat"
	case strings.HasPrefix(branchName, "fix/"):
		return "fix"
	case strings.HasPrefix(branchName, "docs/"):
		return "docs"
	case strings.HasPrefix(branchName, "refactor/"):
		return "refactor"
	case strings.HasPrefix(branchName, "test/"):
		return "test"
	}

	// Check file categories
	switch {
	case len(categories["test"]) > 0 && len(categories["source"]) == 0:
		return "test"
	case len(categories["docs"]) > 0 && len(categories["source"]) == 0:
		return "docs"
	case len(categories["config"]) > 0 && len(categories["source"]) == 0:
		return "chore"
	case len(categories["source"]) > 0:
		for _, f := range categories["source"] {
			if strings.Contains(strings.ToLower(f), "quantum") {
				return "quantum"
			}
		}
		return "feat"
	}

	return "chore"
}

// GenerateScope creates commit scope from changed files
func (ac *AutoCommit) GenerateScope(categories map[string][]string) string {
	scopes := []string{}

	moduleNames := []string{
		"quantum_core", "dll_bridge", "go_runtime", "sycl_accelerator",
		"runtime_engine", "rag_service", "test_orchestrator", "wui_designer",
	}

	var allFiles []string
	for _, list := range categories {
		allFiles = append(allFiles, list...)
	}

	for _, file := range allFiles {
		parts := strings.Split(file, string(filepath.Separator))
		for _, part := range parts {
			for _, mod := range moduleNames {
				if part == mod {
					found := false
					for _, s := range scopes {
						if s == mod {
							found = true
							break
						}
					}
					if !found {
						scopes = append(scopes, mod)
					}
				}
			}
		}

		if strings.Contains(strings.ToLower(file), "hooks") {
			scopes = append(scopes, "hooks")
		}
		if strings.Contains(strings.ToLower(file), "config") {
			scopes = append(scopes, "config")
		}
	}

	// Limit to maximum 2 scopes
	if len(scopes) > 2 {
		scopes = scopes[:2]
	}

	return strings.Join(scopes, ", ")
}

// GenerateCommitMessage creates properly formatted commit message
func (ac *AutoCommit) GenerateCommitMessage(commitType string, scope string, stats map[string]int, categories map[string][]string) string {
	var fileDesc string
	if stats["files"] == 1 {
		var firstFile string
		if len(categories["source"]) > 0 {
			firstFile = categories["source"][0]
		} else {
			firstFile = categories["other"][0]
		}
		fileDesc = fmt.Sprintf("update %s", filepath.Base(firstFile))
	} else {
		fileDesc = fmt.Sprintf("%d files", stats["files"])
	}

	var message string
	switch commitType {
	case "feat":
		if stats["files"] == 1 && len(categories["source"]) > 0 {
			stem := strings.TrimSuffix(filepath.Base(categories["source"][0]), filepath.Ext(categories["source"][0]))
			message = fmt.Sprintf("add %s functionality", stem)
		} else {
			message = fmt.Sprintf("add %s with %d insertions", fileDesc, stats["insertions"])
		}
	case "fix":
		message = fmt.Sprintf("fix issues in %s", fileDesc)
	case "docs":
		message = fmt.Sprintf("update documentation in %s", fileDesc)
	case "refactor":
		message = fmt.Sprintf("refactor %s", fileDesc)
	case "test":
		message = fmt.Sprintf("add/update tests for %s", fileDesc)
	case "chore":
		message = fmt.Sprintf("update %s", fileDesc)
	case "quantum":
		message = fmt.Sprintf("update quantum operations in %s", fileDesc)
	default:
		message = fmt.Sprintf("update %s", fileDesc)
	}

	if scope != "" {
		return fmt.Sprintf("%s(%s): %s", commitType, scope, message)
	}
	return fmt.Sprintf("%s: %s", commitType, message)
}

// ValidateCommitMessage checks message against required format
func (ac *AutoCommit) ValidateCommitMessage(message string) (bool, string) {
	pattern := `^(feat|fix|docs|style|refactor|test|chore|quantum)(\(\w+\))?: .{1,72}$`
	matched, _ := regexp.MatchString(pattern, message)

	if matched {
		return true, "Commit message format is valid"
	}
	return false, fmt.Sprintf("Commit message does not match required format: %s", pattern)
}

// Commit executes full auto commit workflow
func (ac *AutoCommit) Commit(autoPush bool, createPR bool, dryRun bool) CommitResult {
	result := CommitResult{
		Success:   false,
		DryRun:    dryRun,
		PRCreated: false,
	}

	// Verify git repository
	success, _ := ac.RunGit("status")
	if !success {
		result.Error = "Not in a git repository"
		return result
	}

	// Get current branch
	branchName := ac.GetCurrentBranch()
	if branchName == "unknown" {
		result.Error = "Could not determine current branch"
		return result
	}

	// Get staged files
	stagedFiles := ac.GetStagedFiles()
	if len(stagedFiles) == 0 {
		result.Error = "No staged files to commit"
		return result
	}
	result.StagedFiles = stagedFiles

	// Get change statistics
	stats := ac.GetStagedChanges()
	result.Stats = stats

	// Analyze file changes
	categories := ac.AnalyzeFileChanges(stagedFiles)

	// Determine commit type and scope
	commitType := ac.DetermineCommitType(categories, branchName)
	scope := ac.GenerateScope(categories)

	// Generate commit message
	commitMessage := ac.GenerateCommitMessage(commitType, scope, stats, categories)
	result.CommitMessage = commitMessage

	// Validate commit message
	isValid, _ := ac.ValidateCommitMessage(commitMessage)
	if !isValid {
		result.Error = fmt.Sprintf("Generated commit message is invalid: %s", commitMessage)
		return result
	}

	if dryRun {
		result.Success = true
		return result
	}

	// Create commit
	success, output := ac.RunGit("commit", "-m", commitMessage)
	if !success {
		result.Error = fmt.Sprintf("Failed to create commit: %s", output)
		return result
	}

	// Push if requested
	if autoPush {
		success, output := ac.RunGit("push", "-u", "origin", branchName)
		if !success {
			result.Error = fmt.Sprintf("Failed to push: %s", output)
			return result
		}
	}

	// Create PR if requested
	if createPR && autoPush {
		prBody := fmt.Sprintf(`## Auto-Generated PR

This PR was automatically created by the Kanban Auto-Commit workflow.

**Commit Message:** %s

### Changes
- Auto-committed staged changes
- Generated commit message following project conventions

---
*Generated on %s*
`, commitMessage, time.Now().Format(time.RFC3339))

		cmd := exec.Command("gh", "pr", "create",
			"--title", commitMessage,
			"--body", prBody,
			"--base", "develop",
			"--head", branchName)
		cmd.Dir = ac.RepoPath

		out, err := cmd.Output()
		if err == nil {
			result.PRCreated = true
			result.PRURL = strings.TrimSpace(string(out))
		}
	}

	result.Success = true
	return result
}

func main() {
	repoPath := flag.String("repo-path", ".", "Path to git repository")
	autoPush := flag.Bool("auto-push", false, "Push after committing")
	createPR := flag.Bool("create-pr", false, "Create PR after pushing")
	dryRun := flag.Bool("dry-run", false, "Only generate message, don't commit")
	jsonOutput := flag.Bool("json", false, "Output result as JSON")

	flag.Parse()

	ac := NewAutoCommit(*repoPath)
	result := ac.Commit(*autoPush, *createPR, *dryRun)

	if *jsonOutput {
		data, _ := json.MarshalIndent(result, "", "  ")
		fmt.Println(string(data))
	} else {
		if result.Success {
			if result.DryRun {
				fmt.Printf("Would commit with message: %s\n", result.CommitMessage)
				fmt.Printf("Staged files: %d\n", len(result.StagedFiles))
				fmt.Printf("Stats: +%d/-%d\n", result.Stats["insertions"], result.Stats["deletions"])
			} else {
				fmt.Printf("Successfully committed: %s\n", result.CommitMessage)
				if *autoPush {
					fmt.Println("Changes pushed to remote")
				}
				if result.PRCreated {
					fmt.Printf("PR created: %s\n", result.PRURL)
				}
			}
		} else {
			fmt.Fprintf(os.Stderr, "Error: %s\n", result.Error)
			os.Exit(1)
		}
	}
}